fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreOpeningTest.java

```diff
@@ -1033,6 +1037,113 @@ void testVersionChangedFlagAfterOpening() {
         }
     }
 
+    @ParameterizedTest
+    @BooleanSource({"useIndex", "rebuild"})
+    void rebuildBasedOnRecordCounts(boolean useIndex, boolean rebuild) {
+        final RecordMetaDataBuilder builder;
+        if (useIndex) {
+            builder = simpleMetaDataBuilder();
+            addCountIndex(builder, EmptyKeyExpression.EMPTY);
+        } else {
+            builder = addRecordCountKey(simpleMetaDataBuilder(), EmptyKeyExpression.EMPTY);
+        }
+        // Note: RecordMetaDataBuilder.build() does not copy structures, so `initialMetaData` will be modified by
+        // calls below, but not correctly, that's why I put it in a block to ensure it is not reused
+        final int initialVersion;
+        {
+            final RecordMetaData initialMetaData = builder.build();
+            initialVersion = initialMetaData.getVersion();
+            saveRecords(initialMetaData, 0, 100, () -> { }, () -> { });
+        }
+
+        final Index index = new Index("OnNum", "num_value_2");
+        builder.addIndex("MySimpleRecord", index);
+        final RecordMetaData newMetaData = builder.build();
+
+        try (FDBRecordContext context = openContext()) {
+            final Map<String, Long> needRebuildIndexes = new ConcurrentHashMap<>();
+            FDBRecordStore store = getStoreBuilder(context, newMetaData, path).setUserVersionChecker(
+                    new FDBRecordStoreBase.UserVersionChecker() {
+                        @Override
+                        @SuppressWarnings("deprecation")
+                        public CompletableFuture<Integer> checkUserVersion(final int oldUserVersion, final int oldMetaDataVersion, final RecordMetaDataProvider metaData) {
+                            assertEquals(0, oldUserVersion);
+                            assertEquals(initialVersion, oldMetaDataVersion);
+                            assertEquals(newMetaData, metaData);
+                            return CompletableFuture.completedFuture(1);
+                        }
+
+                        @Override
+                        public IndexState needRebuildIndex(final Index indexToRebuild, final long recordCount, final boolean indexOnNewRecordTypes) {
```

Comment:
FWIW, the plan with the user version checker is to remove the `needRebuildIndex` that takes a long and replace it with just the other variant, though we haven't formally deprecated the other method yet. Using this method is probably fine for this PR, and we could address that once the other method is deprecated.

Which, tbh, that class is already a bit of a mess, mainly for backwards compatibility reasons. It would be nice to be able to clean those all up at once

Reply:
ok, updated. I also converted this to an inner class, which I think makes the overall code more readable, and fixed a bug here, where it was putting with `index` rather than `indexToRebuild`.